lib/src/kvs/tx.rs

```diff
@@ -772,21 +782,83 @@ impl Transaction {
 	}
 
 	// Set heartbeat
-	pub async fn set_hb(&mut self, id: Uuid) -> Result<(), Error> {
-		let now = self.clock();
-		let key = crate::key::hb::Hb::new(now.clone(), id.0);
+	pub async fn set_hb(&mut self, ts: Timestamp, id: Uuid) -> Result<(), Error> {
+		let key = crate::key::hb::Hb::new(ts.clone(), id.0);
+		let debug = crate::key::hb::Hb::encode(&key).unwrap();
+		trace!(
+			"Setting heartbeat {} ({:?}) ({:?})",
+			String::from_utf8_lossy(&debug).to_string(),
+			&debug,
+			key
+		);
 		// We do not need to do a read, we always want to overwrite
 		self.put(
 			key,
 			ClusterMembership {
 				name: id.0.to_string(),
-				heartbeat: now,
+				heartbeat: ts,
 			},
 		)
 		.await?;
 		Ok(())
 	}
 
+	// Scans up until the heartbeat timestamp and returns the discovered nodes
+	pub async fn scan_hb(&mut self, time_to: &Timestamp, limit: u32) -> Result<Vec<Hb>, Error> {
```

Comment:
could you add more explanation on the algorithm in here? I feel people will need more context than simply looking at the code for understanding it.